---
title: 'ER gene expression (10X)'
output:
  html_document:
    number_sections: yes
    toc: true
    toc_float: true
    code_folding: show
---

data: 

Hs_in-vitro (NCLCN#261ã€€d6c11, d6c56, d6c86, d6c117)
Hs_in-vivo (AC7w, AC9w, AC10w, YK11w, AC13w, AC16w)


# Set up
```{r}
filename<-"10X_ER gene expression"
orig.ident<-c("AC7w","AC9w","AC10w","YK11w","AC13w","AC16w", "c11", "c56","c86", "c117")
dataset<-c(rep("Hs_in-vivo",6), rep("Hs_in-vitro",4))
mapping<-"cr60GRCh38p12chrxmv1b1"
```

```{r}
if(!dir.exists(filename)){
    dir.create(filename)
}
```

## Load libraries

```{r,message=FALSE,warning=FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(Seurat)
  library(ComplexHeatmap)
  library(data.table)
  library(Biobase)
  library(ggthemes)
})
set.seed(0)
```

```{r}
theme_set(theme(text=element_text(size = 18),
                panel.background = element_rect(fill = "white",color="gray20"),
                panel.grid=element_line(color="gray92"),
                legend.key = element_rect(colour = NA, fill = NA)))
```

```{r}
mytheme<-function(x){
  theme(axis.line.x = element_line(size=1, color="black"),
        axis.line.y = element_line(size=1, color="black"),
        axis.title.y = element_text(size=22,face="bold",angle=90,margin = margin(rep(5,4))),
        axis.title.y.right = element_text(size=22,face="bold",angle=-90,margin = margin(rep(5,4))),
        axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        axis.text.y = element_text(hjust=1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        axis.text.y.right = element_text(hjust=-1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        plot.title = element_text(size=18,hjust = 0),
        legend.text=element_text(size=17, color="black"),
        legend.title=element_text(size=19, color="black"),
        legend.key = element_rect(colour = NA, fill = NA),
        panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_line(size=0.6,linetype="longdash",color="gray80"),
        panel.background = element_blank(),
        strip.background = element_blank(),
        strip.text=element_text(size=22,face="bold"),
        axis.ticks.length=unit(.2, "cm"),
        axis.ticks = element_line(size=0.6),
        panel.spacing=unit(1, "lines")
        )
}
```

```{r}
celltype_pal<-c(colorRampPalette(c("white","#008000"))(4)[4:3],
                colorRampPalette(c("white","#005080"))(4)[4:3],
                colorRampPalette(c("white","red"))(4)[4:2],
                "gold",
                "#b35900",
                "#a020f0"
                ) %>% 
  setNames(c("VEM1","VEM2","EM1","EM2","M1","M2","M3","PLL1","PLL2","ZPD"))
```

## Import data

```{r}
seurat <- readRDS("2_annotation_vivoACYK_c11_56_86_117_res0.8_5000HVGs_exp2in5cells.rds")
seurat<-ScaleData(seurat,features=rownames(seurat))
```

# QC
Evaluate overall gene expression levels to estimate the threshold at which they can be said to be predominantly expressed.

```{r}
pdat<-data.frame(max=rowMax(seurat[["RNA"]]@data %>% as.matrix()),
                 genes=rownames(seurat))

ggplot(pdat,aes(x=max))+
  geom_histogram(binwidth=0.1)+
  geom_vline(xintercept=1)
```

Most of the low-expressed genes have an expression value of less than 1.

# Analysis
## ER gene expression

```{r}
core_ER<-fread("~/ovaries/bulk_RNA-seq/analysis/Hs_expansion/230331_core_ERgene_ag120_c122_intersect.txt",header=FALSE) %>% 
  unlist()
  
list_ERmurase_e<-c("IL22RA1", "BRDT", "LOC102723543", "DAZL", "IL12B", "KHDC1L", "STK31", "ZNF736", "LRRC61", "TUSC1", "C9orf64", "PLAU", "A2M", "LOC105369746", "GTSF1", "PIWIL1", "MEG3", "SPESP1", "SPATA22", "USP6", "TDRD12", "ZNF534", "NLRP9", "ZNF667", "ZNF667-AS1", "LOC107985433", "PRAME", "MAP7D2", "MAGEB2", "MAGEB1", "LDOC1") %>% intersect(core_ER)
list_ERmurase_l<-c("LINC01160", "RHEX", "GCSAML-AS1", "EMX1", "C2CD6", "KRBOX1", "DCAF4L1", "PDCL2", "RBM46", "DDX4", "C5orf47", "HDGFL1", "OOEP", "DDX43", "LOC101928688", "LOC101929128", "PIWIL2", "LOC101926892", "DCAF4L2", "KCNV2", "ANKRD2", "CYP2E1", "CTSF", "GDPD4", "LOC105369655", "SYCP3", "ANHX", "TDRD9", "LOC105376917", "ZNF492", "DSCR8", "CXorf67", "LUZP4", "MAGEC1", "MAGEA11", "MAGEA4") %>% intersect(core_ER)

genes<-c(list_ERmurase_e,list_ERmurase_l)
```

```{r fig.height=8, fig.width=16}
slot="data"
seurat$tmp.cluster<-paste0(seurat$method,"_",seurat$cluster) %>%
  factor(levels=paste0(rep(levels(seurat$method),each=nlevels(seurat$cluster)),"_",rep(levels(seurat$cluster),nlevels(seurat$method))))
cell_cluster="tmp.cluster"

cluster_show=c("in vivo_EM1", "in vivo_EM2", "in vivo_M1", "in vivo_M2", "in vivo_M3", "in vivo_PLL1", "in vivo_PLL2", 
"in vitro_EM1", "in vitro_EM2", "in vitro_M1", "in vitro_M2", "in vitro_M3", "in vitro_PLL1", "in vitro_PLL2") 
cluster_not_show<-levels(seurat@meta.data[,cell_cluster])[(table(seurat@meta.data[,cell_cluster]) %>% as.numeric())<5] %>% intersect(cluster_show)

celltype2_pal<-rep(celltype_pal,2) %>% setNames(.,levels(seurat@meta.data[,cell_cluster]))

group <- seurat@meta.data[,cell_cluster]
mat <- levels(group) %>% 
      lapply(function(y){
        if(y%in%group){
          if (sum(group==y)>1){
            seurat$temp<-group==y
            seu<-AverageExpression(subset(seurat,temp==TRUE),slot=slot,features=genes,assays = "RNA",return.seurat = TRUE)
            seu[["RNA"]]@data
          }else{
            GetAssayData(seurat,slot=slot)[genes,group==y]
          }
        }else{
          rep(NA,length(genes))
        }
      }) %>% do.call(cbind,.) %>% 
      as.data.frame() %>% 
      `colnames<-`(levels(group))
mat_sub<-mat[,cluster_show]
mat_sub[,cluster_not_show]<-NA
```

```{r fig.height=8, fig.width=16}
# order by expression in L cluster
order_heat<-function(mat,group.by=1){
  ord_mat<-apply(mat,1,function(x){
    data.frame(cluster=colnames(mat)[which.max(x)],
               exp=x[colnames(mat)=="in vivo_PLL1"])}) %>% 
    do.call(rbind,.) %>% 
    mutate(order_orig=1:nrow(mat)) %>% 
    mutate(group.by) %>% 
    transform(cluster=factor(cluster,levels=colnames(mat))) %>% 
    arrange(group.by,desc(exp))
  return(ord_mat[,"order_orig"])
}
```

```{r,fig.width=7,fig.height=10}
ord<-order_heat(mat_sub, group.by = factor(c(rep("Early",length(list_ERmurase_e)),rep("Late",length(list_ERmurase_l))),levels=c("Early","Late")))
if (slot=="scale.data"){
  breaksList = seq(-2, 2, by = 0.01)
  col_fun = circlize::colorRamp2(breaksList, colorRampPalette(rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))(length(breaksList)))
}else{
  col_fun = circlize::colorRamp2(c(0, 1, 2, 5), c("white", "#FDEC00", "red", "red4"))
  }

cluster_label<-cluster_show %>% gsub("^.+_","",.) %>% 
                         factor(levels=names(celltype_pal))
ha = HeatmapAnnotation(cluster = cluster_label,
    col = list(cluster=celltype_pal[cluster_label]),
    show_annotation_name =FALSE,
    annotation_legend_param = list(title="",
                                   labels_gp=gpar(fontsize=12))
    )

HM<-Heatmap(mat_sub[ord,],
            name="Expression",
            cluster_columns = F,
            cluster_rows = F,
            use_raster=TRUE,
            col = col_fun,
            column_labels=rep("",ncol(mat_sub)),
            clustering_distance_rows = "euclidean",
            top_annotation = ha,
            row_names_gp = gpar(fontsize = 12,fontface="italic"),
            cluster_row_slices=T,
            column_split=c(rep("in vivo",ncol(mat_sub)/2),rep("in vitro",ncol(mat_sub)/2)) %>% factor(levels=c("in vivo","in vitro")),
            row_split=c(rep("Early ER genes",length(list_ERmurase_e)),rep("Late ER genes",length(list_ERmurase_l))),
            column_title_gp = gpar(fontface="bold.italic",fontsize=18),
            row_dend_width = unit(4, "cm"), 
            row_title_gp = gpar(face="bold",fontsize=18),
            heatmap_legend_param = list(
              title = expression(Expression),
              title_gp=gpar(fontsize=12),
              labels_gp=gpar(fontsize=12),
              legend_height = unit(4, "cm"),
              title_position = "leftcenter-rot"))
pdf(paste0(filename,"/ERgenes_heatmap.pdf"),width=7,height=10)
HM<-draw(HM)
dev.off()

draw(HM)
```

```{r}
mytheme<-function(x){
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.title.y = element_text(size=20,face="bold",angle=90,margin = margin(rep(5,4))),
        axis.title.y.right = element_text(size=20,face="bold",angle=-90,margin = margin(rep(5,4))),
        axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        axis.text.y = element_text(hjust=1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        axis.text.y.right = element_text(hjust=-1,vjust=0.5,size=20,color="black",margin = margin(rep(3,4))),
        plot.title = element_text(size=18,hjust = 0),
        legend.text=element_text(size=17, color="black"),
        legend.title=element_text(size=19, color="black"),
        legend.key = element_rect(colour = NA, fill = NA),
        panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank(), 
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_line(size=0.6,linetype="longdash",color="gray80"),
        panel.background = element_blank(),
        strip.background = element_blank(),
        strip.text.y=element_text(size=20,angle=-90),
        strip.text.x=element_text(size=20,face="italic"),
        axis.ticks.length=unit(.2, "cm"),
        axis.ticks = element_line(size=0.6),
        panel.spacing=unit(1, "lines")
  )
  }
```


```{r,fig.height=7,fig.width=8}
pdat<-mat_sub %>% 
  as.data.frame() %>% 
  rownames_to_column(var="gene") %>% 
  mutate(gene_type=ifelse(gene%in%list_ERmurase_e,"Early ER genes","Late ER genes") %>% factor(levels=c("Early ER genes","Late ER genes"))) %>% 
  tidyr::pivot_longer(-c(gene,gene_type),names_to = "cluster",values_to = "exp") %>% 
  separate(cluster,into=c("method","cluster"),sep="_") %>% 
  transform(method=factor(method,levels=c("in vivo","in vitro")),
            cluster=factor(cluster,levels=levels(seurat$cluster)))
p1<-ggplot(pdat,aes(x=cluster,y=exp,fill=cluster))+
  geom_boxplot(outlier.shape = NA)+
  facet_grid(gene_type~method)+
  xlab("")+ylab("")+
  ylim(0,4)+
  scale_fill_manual(values=celltype_pal)+
  mytheme()+
  ylab(expression(paste("Expression, ",log[e],"(ss100kUMI+1)"))) +
  annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf,size=1.3)+
  annotate("segment", x=-Inf, xend=-Inf, y=-Inf, yend=Inf,size=1.3)
p1

pdf(paste0(filename,"/ERgenes_boxplot.pdf"),width=8,height=7)
p1
dev.off()
```
