---
title: "Untitled"
author: "Yusuke Murase"
date: "2023/1/18"
output: html_document
---

```{r data import}
library(Seurat)
library(Matrix)
library(SeuratObject)
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(SeuratDisk)
library(patchwork)
library(readxl)
library(amap)
library(dendextend)

#load CellRanger output
data_p <- Read10X(data.dir = "[your_directory]/sample_feature_bc_matrix")# WT
seurat_object_p = CreateSeuratObject(counts = data_p$`Gene Expression`, project = "prnt", min.cells = 3, min.features = 200)

data_ko1 <- Read10X(data.dir = "[your_directory]/sample_feature_bc_matrix")# TET1 KO#1
seurat_object_ko1 = CreateSeuratObject(counts = data_ko1$`Gene Expression`, project = "ko1", min.cells = 3, min.features = 200)

data_ko2 <- Read10X(data.dir = "[your_directory]/sample_feature_bc_matrix")# TET1 KO#2
seurat_object_ko2 = CreateSeuratObject(counts = data_ko2$`Gene Expression`, project = "ko2", min.cells = 3, min.features = 200)

cells <- fread("[your_directory]/outs/multi/multiplexing_analysis/assignment_confidence_table.csv",select = c("Barcodes"))
seurat_object_p <- subset(seurat_object_p, cells = cells$Barcodes)
seurat_object_ko1 <- subset(seurat_object_ko1, cells = cells$Barcodes)
seurat_object_ko2 <- subset(seurat_object_ko2, cells = cells$Barcodes)

seurat_object_use <- merge(seurat_object_p, y = c(seurat_object_ko1,seurat_object_ko2), add.cell.ids = c("prnt","ko1","ko2"), project = "T1KO")

saveRDS(seurat_object_use, "seurat_object_use.Rdata")
```

```{r data export for RECODE}
write.table(t(as.matrix(GetAssayData(object = seurat_object_use, slot = "counts"))), 
            'seurat_object_use_T1KO.csv', 
            sep = ',', row.names = T, col.names = T, quote = F)
#run desktop RECODE app with seurat_object_use_T1KO.csv
```

```{r load post RECODE data}
data_RECODE<- read.table("seurat_object_use_T1KO.RECODE.csv", sep =",", header = T, row.names = 1 )

seurat_object_use[["RECODE"]] <- CreateAssayObject(counts = Matrix(t(as.matrix(data_RECODE)), sparse = T))
```

```{r nCount and percent.mt filter}
seurat_object_use[["percent.mt"]] <- PercentageFeatureSet(seurat_object_use, pattern = "^MT-")
VlnPlot(seurat_object_use, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(seurat_object_use, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(seurat_object_use, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

seurat_object_nC10k_nC100k_pmt10 <- subset(seurat_object_use, subset = nCount_RNA >10000 &  nCount_RNA <100000 & percent.mt < 10)
VlnPlot(seurat_object_nC10k_nC100k_pmt10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

```{r preprocessing for scrublet}
#subset object by genotye(CMO tag)
DefaultAssay(seurat_object_nC10k_nC100k_pmt10) <- "RECODE"

prnt <- subset(seurat_object_nC10k_nC100k_pmt10, subset = orig.ident == "prnt")
ko1 <- subset(seurat_object_nC10k_nC100k_pmt10, subset = orig.ident == "ko1")
ko2 <- subset(seurat_object_nC10k_nC100k_pmt10, subset = orig.ident == "ko2")
#run scrublet one by one
#scrublet_prnt.Rmd
#scrublet_ko1.Rmd
#scrublet_ko2.Rmd
```

```{r post-scrublet processing}
#merge 3 post-scrublet seurat objects
SeuratObj_REC_scrublet <- merge(prnt, y = c(ko1,ko2), add.cell.ids = c("prnt","ko1","ko2"), project = "T1KO")
sum(SeuratObj_REC_scrublet@meta.data[["Is_doublets"]])#0, single only
DefaultAssay(SeuratObj_REC_scrublet)
```

```{r data normalization}
SeuratObj_REC_scrublet <- NormalizeData(SeuratObj_REC_scrublet, normalization.method = "LogNormalize", scale.factor = 100000)#10kでなく100k 
SeuratObj_REC_scrublet  <- FindVariableFeatures(SeuratObj_REC_scrublet, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(SeuratObj_REC_scrublet), 10)
options(repr.plot.width = 12, repr.plot.height = 5)
plot1 <- VariableFeaturePlot(SeuratObj_REC_scrublet)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
png("230119_SeuratObj_REC_scrublet_combiplot_100knorm.png", width = 1000)
CombinePlots(plots = list(plot1, plot2))
dev.off()
all.genes <- rownames(SeuratObj_REC_scrublet)
SeuratObj_REC_scrublet <- ScaleData(SeuratObj_REC_scrublet, features = all.genes)
SeuratObj_REC_scrublet <- RunPCA(SeuratObj_REC_scrublet, features = VariableFeatures(object = SeuratObj_REC_scrublet))
p1 <- DimPlot(SeuratObj_REC_scrublet, reduction = "pca", label = TRUE)
p2 <- DimPlot(SeuratObj_REC_scrublet, reduction = "pca", label = TRUE, split.by ="orig.ident")
png("230119_SeuratObj_REC_scrublet_PCA_100knorm.png", width = 1000)
p1+p2
dev.off()
png("230119_SeuratObj_REC_scrublet_elbow_100knorm.png")
ElbowPlot(SeuratObj_REC_scrublet, ndims = 50)
dev.off()
```

```{r Fig.4a}
Idents(SeuratObj_REC_scrublet) <- "orig.ident"
DimPlot(SeuratObj_REC_scrublet, reduction = "umap")

layer_order <- fct_inorder(rev(c("WT","KO#1","KO#2","other")))

Seurat_WT <- RenameIdents(SeuratObj_REC_scrublet,"prnt" = layer_order[4], "ko1" = layer_order[1], "ko2" = layer_order[1])
Seurat_WT@active.ident
pW <- DimPlot(Seurat_WT, reduction = "umap", order = T)+
  scale_color_manual(values = c("gray","red"))+
  theme_void()+
  theme(panel.background = element_rect(fill = "white", colour = "black", size = 1.0))
Seurat_KO1 <- RenameIdents(SeuratObj_REC_scrublet,"prnt" = layer_order[1], "ko1" = layer_order[3], "ko2" = layer_order[1])
Seurat_KO1@active.ident
pK1 <- DimPlot(Seurat_KO1, reduction = "umap", order = T)+
  scale_color_manual(values = c("gray", "red"))+
  theme_void()+
  theme(panel.background = element_rect(fill = "white", colour = "black", size = 1.0))
Seurat_KO2 <- RenameIdents(SeuratObj_REC_scrublet,"prnt" = layer_order[1], "ko1" = layer_order[1], "ko2" = layer_order[2])
Seurat_KO2@active.ident
pK2 <- DimPlot(Seurat_KO2, reduction = "umap",order = T)+
  scale_color_manual(values = c("gray", "red"))+
  theme_void()+
  theme(panel.background = element_rect(fill = "white", colour = "black", size = 1.0))

ggsave("231221_DP_WTonly.pdf", pW+NoLegend(), width = 1.53*3, height = 1.1*3 )
ggsave("231221_DP_KO1only.pdf", pK1+NoLegend(), width = 1.53*3, height = 1.1*3 )
ggsave("231221_DP_KO2only.pdf", pK2+NoLegend(), width = 1.53*3, height = 1.1*3 )
```

```{r Fig.4b}
SeuratObj_REC_scrublet <- FindNeighbors(SeuratObj_REC_scrublet, dims = 1:22)
SeuratObj_REC_scrublet <- FindClusters(SeuratObj_REC_scrublet, resolution = 0.5)
SeuratObj_REC_scrublet <- RunUMAP(SeuratObj_REC_scrublet, dims = 1:22)

saveRDS(SeuratObj_REC_scrublet, "SeuratObj_REC_scrublet_dim22_reso05.Rdata")
SaveH5Seurat(SeuratObj_REC_scrublet,filename =  "SeuratObj_REC_scrublet_dim22_reso05.h5Seurat")

SeuratObj_REC_scrublet <- readRDS("[your directory]/SeuratObj_REC_scrublet_dim22_reso05.Rdata")

Idents(SeuratObj_REC_scrublet)

new.cluster.ids <- c("cluster_1", #0
                     "cluster_4", #1
                     "cluster_5", #2 
                     "cluster_2", #3
                     "cluster_6", #4 
                     "cluster_8", #5
                     "cluster_9", #6
                     "cluster_7", #7
                     "cluster_3", #8
                     "cluster_10", #9
                     "cluster_11" #11
)
new.cluster.ids <- factor(new.cluster.ids, levels = c("cluster_1","cluster_2","cluster_3","cluster_4","cluster_5","cluster_6","cluster_7","cluster_8","cluster_9","cluster_10","cluster_11"))
SetIdent(SeuratObj_REC_scrublet, value = "seurat_clusters")
Idents(SeuratObj_REC_scrublet) %>% table()
SeuratObj_REC_scrublet <- RenameIdents(SeuratObj_REC_scrublet, 
                                       "0" = new.cluster.ids[1],
                                       "1" = new.cluster.ids[2],
                                       "2" = new.cluster.ids[3],
                                       "3" = new.cluster.ids[4],
                                       "4" = new.cluster.ids[5],
                                       "5" = new.cluster.ids[6],
                                       "6" = new.cluster.ids[7],
                                       "7" = new.cluster.ids[8],
                                       "8" = new.cluster.ids[9],
                                       "9" = new.cluster.ids[10],
                                       "10" = new.cluster.ids[11])

#Draw plot by Dimplot() or FeaturePlot() For Fig.4

```
```{r cluster compariason Extended Data Fig11a,e}
hoge <- SplitObject(SeuratObj_REC_scrublet,split.by = "orig.ident")
prnt_composition <- hoge$prnt@meta.data %>% .$seurat_clusters %>% table %>% data.frame() %>% rename("cl" = ".", "prnt" = "Freq")
ko1_composition <-hoge$ko1@meta.data %>% .$seurat_clusters %>% table %>% data.frame() %>% rename("cl" = ".", "ko1" = "Freq")
ko2_composition <-hoge$ko2@meta.data %>% .$seurat_clusters %>% table %>% data.frame() %>% rename("cl" = ".", "ko2" = "Freq")
cl_col <-c(rev(brewer.pal(3,"Oranges")),rev(brewer.pal(8,"Greens")))
composition <- left_join(by = "cl", left_join(by = "cl",prnt_composition,ko1_composition), ko2_composition) %>% 
  mutate("cl_name" = c("cluster_1", #0
                       "cluster_4", #1
                       "cluster_5", #2 
                       "cluster_2", #3
                       "cluster_6", #4 
                       "cluster_8", #5
                       "cluster_9", #6
                       "cluster_7", #7
                       "cluster_3", #8
                       "cluster_10", #9
                       "cluster_11" #11
),
         fct.cl_name = factor(cl_name, levels = c("cluster_1","cluster_2","cluster_3","cluster_4","cluster_5","cluster_6","cluster_7","cluster_8","cluster_9","cluster_10","cluster_11")))


p <- composition %>% select(-c(cl,cl_name)) %>% pivot_longer(-fct.cl_name) %>% ggplot(aes(x = fct(name),y = value,fill = fct.cl_name))+
  geom_col(position = "fill", color = "black")+
  scale_fill_manual(values =  cl_col)+
  theme_classic()+
   theme(plot.background = element_blank(),
        #panel.grid.minor = element_blank(),
        #panel.grid.major = element_blank(),
        panel.background = element_blank(),
        #axis.line = element_blank(),
        #axis.ticks = element_blank(),
        strip.background = element_rect(fill = "white", colour = "white"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text = element_text(colour = "black"),
        legend.text = element_text(colour = "black"),
        legend.title = element_blank(),
        axis.title = element_blank())

ggsave(paste0(Sys.Date(),"_compotision.pdf"), p, dpi = 300, height = 3, width =2.75)

hoge <- SplitObject(SeuratObj_REC_scrublet,split.by = "celltype")
cls <- list()
cl <- c("cluster_1","cluster_2","cluster_3","cluster_4","cluster_5","cluster_6","cluster_7","cluster_8","cluster_9","cluster_10","cluster_11")
cls[[1]] <- hoge$cluster_1@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_1" = "Freq")
cls[[2]] <- hoge$cluster_2@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_2" = "Freq")
cls[[3]] <- hoge$cluster_3@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_3" = "Freq")
cls[[4]] <- hoge$cluster_4@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_4" = "Freq")
cls[[5]] <- hoge$cluster_5@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_5" = "Freq")
cls[[6]] <- hoge$cluster_6@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_6" = "Freq")
cls[[7]] <- hoge$cluster_7@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_7" = "Freq")
cls[[8]] <- hoge$cluster_8@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_8" = "Freq")
cls[[9]] <- hoge$cluster_9@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_9" = "Freq")
cls[[10]] <- hoge$cluster_10@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_10" = "Freq")
cls[[11]] <- hoge$cluster_11@meta.data %>% .$orig.ident %>% table %>% data.frame() %>% rename("geno" = ".", "cluster_11" = "Freq")
　
geno_comp <- cls[[1]] %>% left_join(by = "geno", cls[[2]]) %>% left_join(by = "geno", cls[[3]]) %>% left_join(by = "geno", cls[[4]]) %>% left_join(by = "geno", cls[[5]]) %>% left_join(by = "geno", cls[[6]]) %>% left_join(by = "geno", cls[[7]])%>% left_join(by = "geno", cls[[8]])%>% left_join(by = "geno", cls[[9]])%>% left_join(by = "geno", cls[[10]])%>% left_join(by = "geno", cls[[11]]) 

geno_comp$cluster_1 <-  geno_comp$cluster_1   %>%  replace_na(0)
geno_comp$cluster_2 <-  geno_comp$cluster_2   %>%  replace_na(0)
geno_comp$cluster_3 <-  geno_comp$cluster_3   %>%  replace_na(0)
geno_comp$cluster_4 <-  geno_comp$cluster_4   %>%  replace_na(0)
geno_comp$cluster_5 <-  geno_comp$cluster_5   %>%  replace_na(0)
geno_comp$cluster_6 <-  geno_comp$cluster_6   %>%  replace_na(0)
geno_comp$cluster_7 <-  geno_comp$cluster_7   %>%  replace_na(0)
geno_comp$cluster_8 <-  geno_comp$cluster_8   %>%  replace_na(0)
geno_comp$cluster_9 <-  geno_comp$cluster_9   %>%  replace_na(0)
geno_comp$cluster_10 <- geno_comp$cluster_10  %>%  replace_na(0)
geno_col <-  c(rgb(232/256,	125/256,	113/256	),
    rgb(83/256,	181/256,	75/256),
    rgb(109/256, 157/256, 248/256)
    )

p <- geno_comp %>% 
  mutate("genotype" = c("KO#1", "KO#2","parent"),
         fct.geno = factor(genotype, levels = c("parent","KO#1", "KO#2"))) %>% 
  select(-c("geno","genotype")) %>% pivot_longer(-fct.geno) %>% 
  ggplot(aes(x = fct(name),y = value, fill = fct.geno))+
  geom_col(#position = "fill", 
           color = "black")+
  scale_fill_manual(values =  geno_col)+
  theme_classic()+
   theme(plot.background = element_blank(),
        #panel.grid.minor = element_blank(),
        #panel.grid.major = element_blank(),
        panel.background = element_blank(),
        #axis.line = element_blank(),
        #axis.ticks = element_blank(),
        strip.background = element_rect(fill = "white", colour = "white"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text = element_text(colour = "black"),
        legend.text = element_text(colour = "black"),
        legend.title = element_blank(),
        axis.title = element_blank())
  
ggsave(paste0(Sys.Date(),"_geno_compotision_wide.pdf"), p, dpi = 300, height = 2.2, width =3.5)

write.table(composition, "240226_composition_T1KO.txt", row.names = F)

```
```{r cell cycle scoring Extebded Data Fig11b}
SeuratObj_REC_scrublet_ccr <- SeuratObj_REC_scrublet
SeuratObj_REC_scrublet_ccr <- CellCycleScoring(SeuratObj_REC_scrublet_ccr, 
                                               s.features   = UpdateSymbolList(symbols = cc.genes$s.genes), 
                                               g2m.features = UpdateSymbolList(symbols = cc.genes$g2m.genes), set.ident = TRUE)

SeuratObj_REC_scrublet_ccr$Phase <- factor(SeuratObj_REC_scrublet_ccr$Phase, levels = c("G1","S","G2M")) 
 
#ggsave(paste0(Sys.Date(),"_phase_DimpPlot.tiff"), p, dpi = 300, height = 3.5, width = 3.5)

SeuratObj_REC_scrublet_ccr@active.ident
hoge <- SplitObject(SeuratObj_REC_scrublet_ccr,split.by = "celltype")

cls <- list()
cl <- c("cluster_1","cluster_2","cluster_3","cluster_4","cluster_5","cluster_6","cluster_7","cluster_8","cluster_9","cluster_10","cluster_11")
cls[[1]] <- hoge$cluster_1@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_1" = "Freq")
cls[[2]] <- hoge$cluster_2@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_2" = "Freq")
cls[[3]] <- hoge$cluster_3@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_3" = "Freq")
cls[[4]] <- hoge$cluster_4@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_4" = "Freq")
cls[[5]] <- hoge$cluster_5@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_5" = "Freq")
cls[[6]] <- hoge$cluster_6@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_6" = "Freq")
cls[[7]] <- hoge$cluster_7@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_7" = "Freq")
cls[[8]] <- hoge$cluster_8@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_8" = "Freq")
cls[[9]] <- hoge$cluster_9@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_9" = "Freq")
cls[[10]] <- hoge$cluster_10@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_10" = "Freq")
cls[[11]] <- hoge$cluster_11@meta.data %>% .$Phase %>% table %>% data.frame() %>% rename("phase" = ".", "cluster_11" = "Freq")
　
phase_comp <- cls[[1]] %>% left_join(by = "phase", cls[[2]]) %>% left_join(by = "phase", cls[[3]]) %>% left_join(by = "phase", cls[[4]]) %>% left_join(by = "phase", cls[[5]]) %>% left_join(by = "phase", cls[[6]]) %>% left_join(by = "phase", cls[[7]])%>% left_join(by = "phase", cls[[8]])%>% left_join(by = "phase", cls[[9]])%>% left_join(by = "phase", cls[[10]])%>% left_join(by = "phase", cls[[11]]) 

#NA to zero
phase_comp$cluster_2 <- phase_comp$cluster_2  %>%  replace_na(0)
phase_comp$cluster_10 <- phase_comp$cluster_10  %>%  replace_na(0)
phase_comp$cluster_11 <- phase_comp$cluster_11  %>%  replace_na(0)
phase_col <-  c("#0066E7",
    "#FFDA13",
    "#D4070F"
    )

p <- phase_comp %>% 
  mutate("phasetype" = c("G1", "S","G2M"),
         fct.phase = factor(phasetype, levels = c("G1","G2M","S"))) %>% 
  select(-c("phase","phasetype")) %>% 
  pivot_longer(-fct.phase) %>% 
  ggplot(aes(x = fct(name),y = value, fill = fct.phase))+
  geom_col(position = "fill", 
           color = "black")+
  scale_fill_manual(values =  phase_col)+
  theme_classic()+
   theme(plot.background = element_blank(),
        #panel.grid.minor = element_blank(),
        #panel.grid.major = element_blank(),
        panel.background = element_blank(),
        #axis.line = element_blank(),
        #axis.ticks = element_blank(),
        strip.background = element_rect(fill = "white", colour = "white"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text = element_text(colour = "black"),
        legend.text = element_text(colour = "black"),
        legend.title = element_blank(),
        axis.title = element_blank())
  
#ggsave(paste0(Sys.Date(),"_phase_compotision.pdf"), p, dpi = 300, height = 2.5, width =3.5)
saveRDS(SeuratObj_REC_scrublet, "SeuratObj_T1KO.rds")
```
```{r cluster correlation Extended Data Fig11c}
av.SeuratObj_REC_scrublet <- AverageExpression(SeuratObj_REC_scrublet, return.seurat = T)
av.SeuratObj_REC_scrublet@assays$RECODE@data#log-norm average expression value
av.var.gene <- av.SeuratObj_REC_scrublet@assays$RECODE@data %>% as.data.frame() %>% rownames_to_column() %>% filter(rowname %in% VariableFeatures(SeuratObj_REC_scrublet)) %>% column_to_rownames("rowname") %>% as.matrix()
cor.exp <- as.data.frame(cor(av.var.gene, method = "spearman"))
cor.exp$x <- rownames(cor.exp)

dend <- hclust(d = Dist(t(av.var.gene), method = "euclidean"), method = "ward.D2")
plot(dend, hang = -1)

pdf(paste0(Sys.Date(),"_UHC_EUC_WD2_top2000.pdf"),dend ,height = 4)
dend %>% rotate(c(2,3,1,8,9,7,10,11,6,5,4)) %>% plot(hang = -1)
dev.off()

p <-cor.exp %>%  pivot_longer(-x) %>% mutate(fct.x = fct_inorder(x),
                                         fct.name = factor(name, levels = rev(c("cluster_1",
                                                      "cluster_2",
                                                      "cluster_3",
                                                      "cluster_4",
                                                      "cluster_5",
                                                      "cluster_6",
                                                      "cluster_7",
                                                      "cluster_8",
                                                      "cluster_9",
                                                      "cluster_10",
                                                      "cluster_11")))) %>% 
  ggplot(aes(x = fct.x, y = fct.name, fill = value))+
  scale_fill_viridis_c(option = "A")+
  geom_tile()+
  theme_bw()+
  theme(plot.background = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        strip.background = element_rect(fill = "white", colour = "white"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text = element_text(colour = "black"),
        legend.text = element_text(colour = "black"),
        axis.title = element_blank())
ggsave(paste0(Sys.Date(),"_CORRHM_dim22reso05_top2000.pdf"), p, dpi = 300, height = 3, width =4)
```
```{r data export PAGA analysis Extended Data Fig11d}
SeuratObj_REC_scrublet$celltype <- as.character(Idents(SeuratObj_REC_scrublet))
SaveH5Seurat(SeuratObj_REC_scrublet, filename = "SeuratObj_REC_scrublet_paga_input.h5Seurat")
Convert("SeuratObj_REC_scrublet_paga_input.h5Seurat", dest = "h5ad")

```
```{r marker gene expression Extended Data Extended Data Fig.11e}
genes  <- c(
  "NANOG","POU5F1", #PGC
  "TFAP2C","SOX17","PRDM1", #PGC
  "HAND1","GATA3","TFAP2A","CDX2",#AME and TE
  "SDC1","ELF5","KRT7","KRT19",#Trophoblast
  "ITGB4","NR2F2",#Trophoblast 
  "ISL1","BMP4","WNT6","GABRP","VTCN1", #AME
  "FOXF1","VIM", #EXMC
  "PECAM1","SOX18","CLDN5",
  "PDGFRA","PITX2",
  "AFP","TTR",
  "MKI67", "TOP2A")


p <- DotPlot(SeuratObj_REC_scrublet, features = genes)+
  scale_color_viridis_c()+
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size = 10),
        plot.margin = margin(0,0,0,0,unit = "cm"))+
  coord_flip()

```
